awx/main/management/commands/run_callback_receiver.py

```diff
@@ -161,14 +162,35 @@ def _save_event_data():
                         break
 
                 if body.get('event') == 'EOF':
-                    # EOF events are sent when stdout for the running task is
-                    # closed. don't actually persist them to the database; we
-                    # just use them to report `summary` websocket events as an
-                    # approximation for when a job is "done"
-                    emit_channel_notification(
-                        'jobs-summary',
-                        dict(group_name='jobs', unified_job_id=job_identifier)
-                    )
+                    try:
+                        logger.info('Event processing is finished for Job {}, sending notifications'.format(job_identifier))
+                        # EOF events are sent when stdout for the running task is
+                        # closed. don't actually persist them to the database; we
+                        # just use them to report `summary` websocket events as an
+                        # approximation for when a job is "done"
+                        emit_channel_notification(
+                            'jobs-summary',
+                            dict(group_name='jobs', unified_job_id=job_identifier)
+                        )
+                        # Additionally, when we've processed all events, we should
+                        # have all the data we need to send out success/failure
+                        # notification templates
+                        uj = UnifiedJob.objects.get(pk=job_identifier)
+                        if isinstance(uj, JobNotificationMixin):
+                            retries = 0
+                            while retries < 5:
+                                if uj.finished:
+                                    uj.send_notification_templates('succeeded' if uj.status == 'successful' else 'failed')
+                                    break
+                                else:
+                                    # wait a few seconds to avoid a race where the
+                                    # events are persisted _before_ the UJ.status
+                                    # changes from running -> successful
+                                    retries += 1
+                                    time.sleep(1)
+                                    uj = UnifiedJob.objects.get(pk=job_identifier)
```

Comment:
@chrismeyersfsu was very interested in this problem. I expected that the job will be marked as finished before the final event comes in sometimes. They might alternate randomly. Solutions we discussed were scheduling a later task, but that creates all kinds of problems with reconciliation.

This is a simply retry 5 x 1 seconds. No matter what solution we have, some of them will be lost, and I don't want to spend a lot of time on the 0.001% of realistic cases. This approach simply ties up a callback worker for 5 seconds. I think that sounds fine.

Reply:
Yep, I saw situations where it went back and forth - sometimes the job _was_ finished, and sometimes it wasn't.

> This approach simply ties up a callback worker for 5 seconds. I think that sounds fine.
Well, one of the workers for _up to_ 5 seconds every time a job finishes.

Reply:
If he gives this 👍 then I do too